## Introduction
### What are python dictionaries 
A dictionary in python is a mapping object that maps keys to values, where the keys are unique within a dictionary and the values can hold any arbitrary value. In addition to being unique, keys are also required to be hashable. An object is said to be hashable if it has a hash value (implemented by a `__hash__()` method) that doesn’t change during the object’s lifetime. Most commonly, we use immutable data types, such as strings, integers, and tuples (only if they contain similarly immutable types) as dictionary keys. A dictionary’s data is always enclosed by a pair of curly braces `{ }`.
Typically dictionaries look like this.

In [1]:
my_dict = {"first_name":"John","last_name":"Snow","age":16,"gender":"Male"}

We have created a dictionary named `my_dict` where each key-value pair is separated by a full colon, with the keys as:
- `first_name`
- `last_name`
- `age`
- `gender`

The values from `my_dict` are:
- `John`
- `Snow`
- `16`
- `Male`


### How different are dictionaries from other common data structures in python
Unlike sequenced data types like lists and tuples where indexing is achieved using the positional indices - which is a range of integers, dictionaries are indexed using their keys. As a result individual values can be accessed using these keys.
### How similar are dictionaries to other data structures in other languages
Like lists, tuples and sets, dictionaries are iterable i.e they can be looped through. Dictionaries are also mutable since they can be modified in place, just like lists and sets.
### Use cases of dictionaries
Typically dictionaries are used to store associative data, i.e data that is related. Examples include:
 - The attributes of an object.
 - A row of data from an `SQL` table.
In this particular scenario, we will be using using dictionaries to store job listing details from [Kaggle](https://www.kaggle.com/ardenn/brighter-monday-job-listings/data)

## Dictionary Operations
### Creating a Dictionary
- To create an empty dictionary, we use a pair of curly braces with nothing between them


In [2]:
empty_dict = {}

In the line above, we have created an empty dictionary named `empty_dict`.

- To create a dictionary with items,we use a pair of curly braces with the key-value pairs. We will create a dictionary to represent the second row of data in the `jobs.csv` file.

In [3]:
job1 = {"title":"Production Manager","location":"Rest of Kenya","job_type":"Full Time",
             "employer":"The African Talent Company (TATC)","category":"Farming"}
job1

{'title': 'Production Manager',
 'location': 'Rest of Kenya',
 'job_type': 'Full Time',
 'employer': 'The African Talent Company (TATC)',
 'category': 'Farming'}

We just created a dictionary with the keys `title`,` location`, `job_type`, `employer`, `category` and assigned it to the variable `job1`.

- To create a dictionary using the `dict()` constructor, we pass it a sequence of key-value pairs. We could also pass the constructor named arguments. We will create a dictionary to represent the third row of data in the `jobs.csv` file, using both of these methods.

In [4]:
#create an empty dictionary
empty_property = dict()

#create dictionary using a list of key-value tuples
job2 = dict([("title","Marketing & Business Development Manager"),("location","Mombasa"),\
("job_type","Full Time"),\
("employer","KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)"),\
("category","Marketing & Communications")])
job2

{'title': 'Marketing & Business Development Manager',
 'location': 'Mombasa',
 'job_type': 'Full Time',
 'employer': 'KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)',
 'category': 'Marketing & Communications'}

In the case above, we passed a sequence, in this case a list of key-value pairs (tuples) to the `dict()` constructor to create our dictionary, and assigned it to the variable `job2`.

In [5]:
#Using keyword arguments
dict(title="Marketing & Business Development Manager",location="Mombasa",job_type="Full Time",
     employer="KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",
     category="Marketing & Communications")

{'title': 'Marketing & Business Development Manager',
 'location': 'Mombasa',
 'job_type': 'Full Time',
 'employer': 'KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)',
 'category': 'Marketing & Communications'}

We created a dictionary with the keys as the argument names, and the values as the argument values. However, this method is only suitable when our keys are just simple strings.

### Accessing Items
- As we mentioned earlier on, dictionaries are indexed using their keys.To access a particular value in a dictionary we use the indexing operator(key inside square brackets). 
- Similarly, we can use the `get()` method of dictionaries to get the value associated with a particular key. We will use this method to access the `title` from `job2`. 

In [6]:
#Using key indexing
job2["title"] #return 'Marketing & Business Development Manager'

#Using get() method
job2.get("title") #return 'Marketing & Business Development Manager'
job2.get("salary") #return None

#Passing a second argument to get()
job2.get("salary", 5000) #return 5000

5000

- In the example above we use indexing to access the `title` from `job2`. Similarly we use the `get()` dictionary method to access the `title` and `salary` values. 
- In the event that we attempt to access a key that doesn't exist, we will get a `KeyError`. To counter this, we use the `get()` method of a dictionary, and pass it the key name. The `get()` method also takes in an optional second argument to be return if the key in the first argument is not found. In the event that we do not want to use the `get()` method, we could check a dictionary for the availability of a certain key using the `in` operator.
```
"salary" in job2 #returns False
"title" in job2 #returns True
```
- However, `job2` doesn't have a `salary` key and as such the value is `None`. However, when we add a second argument, to the `get()` method, the return value is now `5000` instead of `None`.

### Adding and Modifying Items
Dictionaries can be modified directly using the keys or using the `update()` method. The `update()` method take in a dictionary with the key-value pairs to be modified or added. We will add a new item (salary) to `job2` with a value of 10000, modify the `job_type` to be "Part time", update the `salary` to 20000 an finally update the dictionary to include a new item (available) with a value of `True`.

In [7]:
# Adding a new entry for salary using the index
job2["salary"] = 10000

# Modifying the entry for job_type using the index
job2["job_type"] = "Part time"

#Modifying the salary entry using update
job2.update({"salary":20000})

#Adding the available entry using update
job2.update({"available":True})

To add a new entry to a dictionary, we use a similar syntax as indexing. If the index exists, then the value will be modified, however, if the key doesn't exist, a new entry will be created, with the specified key and value. 
- In the first example we assigned a value of 10000 to the `salary` index, but since that index, doesn't exist, a new entry is created, with that value. 
- For the second scenario, the `job_type` key exists, and as such the value is modified to "Part time".
- In the third example we use the `update()` method to change the `salary` value to 20000, since `salary` is already a key in the dictionary.
- Finally, we apply update to the dictionary,and since the `available` item doesn't exist, a new item is created with a key of `available` and value of `True`.

### Removing Items
 We will be removing the just created `salary` item from `job2`. We will also remove everything fro  `job1`.

In [8]:
del job2["salary"]
del job2["available"]
print(job2) #return a dictionary without 'salary' and 'available' items

job1.clear()
print(job1) #return an empty dictionary

del job1

{'title': 'Marketing & Business Development Manager', 'location': 'Mombasa', 'job_type': 'Part time', 'employer': 'KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)', 'category': 'Marketing & Communications'}
{}


- To remove the item associated with the `salary` and `available` keys, we use the `del` keyword as in the case 1 above. If we go ahead and print `job2` we get a dictionary without the `salary`,`available` items.
- To remove all items in a dictionary we use the `clear()` method as in case 2`job1` above. So if we print `job1` we get an empty dictionary.
- If we don't need a dictionary entirely we can use the `del` keyword on the dictionary itself to delete it. Now if we print `job1` we get a `NameError` since `job1` is no longer defined.

### Iterating Through Dictionaries
Since dictionaries are iterable, we can iterate through them in 3 different ways:
 - dict.values() - this returns an iterable of the dictionary's values.
 - dict.keys() - this returns an iterable of the dictionary's keys.
 - dict.items() - this returns an iterable of the dictionary's (key,value) pairs.
 
We will use a `for-loop` to iterate over the `job2` dictionary for all the three methods.

In [9]:
#Using values()
for val in job2.values():
    print(val) #prints the values of job2
    
#Using keys()
for key in job2.keys():
    print(key) #prints the keys of job2
    
#Using items()
for key,val in job2.items():
    print(key, val) #prints the keys and values of job2
    
for _ in job2:
    print(_) #prints the keys of job2

Marketing & Business Development Manager
Mombasa
Part time
KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)
Marketing & Communications
title
location
job_type
employer
category
title Marketing & Business Development Manager
location Mombasa
job_type Part time
employer KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)
category Marketing & Communications
title
location
job_type
employer
category


- In scenario 1 above, we loop through the iterable `job2.values()` and print out the value during each iteration.
- In the second scenario, we iterate through `job2.keys()` while printing out the key. This is the default behaviour when we loop through the entire dictionary, without specifying whether we want the keys, values or items. So scenario 4 gives us a similar result to scenario 2.
- The third scenario allows us to simultaneously loop through the keys and values. We therefore include both key, and value in our for loop constructor since `job.items()` yields a tuple of key and value during each iteration. Our loop therefore prints out the pair at each step.

### Sorting Dictionaries
Borrowing from our description of dictionaries earlier, this data type in meant to be unordered, and doesn't come with the sorting functionality. Calling the `sorted()` function and passing it a dictionary only returns a list of the keys in a sorted order. This is so since the default way to iterate through dictionaries is over its keys.
If we use the `items()` iterable we could sort the items of our dictionary as we please. However, this doesn't give us our original dictionary, but an array of key-value tuples in a sorted order.

In [10]:
#Using sorted() to sort a dictionary's items on the keys
sorted(job2.items(),key=lambda item:item[0])


[('category', 'Marketing & Communications'),
 ('employer',
  'KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)'),
 ('job_type', 'Part time'),
 ('location', 'Mombasa'),
 ('title', 'Marketing & Business Development Manager')]

In this example we use python's inbuilt `sorted()` function which takes in an iterable, in this case our dictionary's items. In the key argument of the `sorted()` function, we specify a function that takes in a key-value pair and returns the key(index 0 of the tuple). This argument instructs `sorted()` to use the value at index 0 for sorting. Similarly, to sort the key-value pairs by the value, we use index 1 instead of index 0.

### Other Dictionary Methods
Dictionaries in python have several other methods, that could be used on demand. To read up further on these, please consult the [python documentation](https://docs.python.org/3/library/stdtypes.html#typesmapping). Here are some other useful methods:
- `pop(key,default)` - deletes the key `key` and returns it, or returns an optional `default` when the key doesn't exist.
- `len(d)` - returns the number of items in a dictionary `d`.
- `copy()` - returns a shallow copy of the original. This shallow copy has similar references to the original, not copies of the original's items. 
- `setdefault(key,default)` - returns the value of `key` if in the dictionary, or sets the new key with an optional `default` as its value and returns the value.

## Speeding Up your Code with Dictionaries
### Dictionary Unpacking
Dictionary unpacking is the ability to desctructure a python dictionary into individual keyword arguments with values. This is especially useful for cases that involve supplying multiple keyword arguments for example in function calls. A useful method is to use the iterable unpacking operator (`**`) to unpack dictionaries.

In [11]:
#Creating a job object
from jobs import Job
new_job = Job(**job2)
print(new_job)

Marketing & Business Development Manager


In our example above, we define a new `Job` class, with the `title`, `location`,`job_type`,`employer`, and `category` attributes. We also define the `__str__()` method which allows us to return title if we print the `Job` object. To instantiate a new `Job`, traditionally, we would need to pass in all the required arguments, but in this case, we just passed in a dictionary with the `**` operator before it. The operator allows us to unpacka dictionary in to an arbitrary number of named arguments, hence speeding up our code.

## Downside Of Using Dictionaries
Compared to lists and tuples, dictionaries take up more space in memory, since they need to store both the key and value, as opposed to just values. Therefore, dictionaries should only be used in cases where we have associative data, that would lose meaning if stored in lists. Python dictionaries however are well-designed such that we can find a value instantly without necessarily having to search through the entire dictionary.

### When not to use dictionaries
- Since dictionaries are unordered, it would not be wise for us to use it to store data that is strictly arranged.
- Dictionaries are also mutable, and as such not suitable for storing data than shouldn't be modified in place.

### How not to use dictionaries
Iterating through a dictionary to search for a specific value is a highly ineffiecient operation. It is however much faster, and cleaner to use the `get()` method.

In [12]:
# How not to search for a value and return it
key_i_need = "location"
value = ""
for key in job2:
    if key == key_i_need:
        value = job2[key]
        
# How to search efficiently
value = job2.get("location")

In the snippet above, we have a variable `key_i_need` containing the key we want to search for inside the dictionary. We then use a for loop to traverse the dictionary, and comparing the key at each step with our variable. If the two match, then we assign the variable `value` to that key's value. This is the wrong approach. We should use the `get()` method on the dictionary, and pass it the desired key instead.

## Performance Tradeoffs
In this section, we will be comparing the space-time tradeoffs between dictionaries and objects. We will use the `timeit` and `getsizeof` modules to determine the time and space aspects. The operations we will be testing are:
- Accessing an entry in a dictionary vs accessing a field in a object
- Adding a new entry to a dictionary vs adding a new field to an object

In terms of space, we will be testing the following operations:
- Size of object classes
- Size of dictionary and object after adding one entry and one field respectively

### Speed Tests
#### Accessing a single entry from a dictionary vs accessing a field in a object

In [13]:
import timeit
#Accessing entry using indexing operator
timeit.timeit('string="Random String"+job3["employer"]',setup='job3 = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}',number=10000)

0.0028828360000261455

In [14]:
#Accessing entry using get() method
timeit.timeit('string="Random String"+job3.get("employer")',setup='job3 = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}',number=10000)

0.0017675090002740035

In [15]:
#Accessing object field using namespace operator
timeit.timeit('string="Random String"+job3.employer',setup='from jobs import Job; job3 = Job("Loans Manager","Mombasa","Full Time",\
        "KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
        "Accounting & Auditing")',number=10000)

0.003938941000342311

In the three tests above, we are looking at the times it takes for each of the three operations to access an entry/ field. Based on the output, the operations can be ranked as:
- Fastest - acessing an entry from a dictionary using indexing operator
- Faster - accessing an entry from a dictionary using `get()` method
- Slowest - accessing a field from an object using the namespace operator

#### Adding a new entry to a dictionary vs adding a new field to an object

In [16]:
#Adding a new entry using indexing and assignment operators
timeit.timeit('job3["salary"]=50000',setup='job3 = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}',number=10000)

0.0019071569995503523

In [17]:
#Adding a new entry using update() method
timeit.timeit('job3.update({"salary":50000})',setup='job3 = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}',number=10000)

0.0021831770000062534

In [18]:
#Adding a new field to the object
timeit.timeit('job3.salary = 50000',setup='from jobs import Job; job3 = Job("Loans Manager","Mombasa",\
"Full Time",\"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"Accounting & Auditing")',number=10000)

0.002361537000069802

In this second round of tests, we are measuring the time it takes to add a new entry to an existing dictionary and the time it takes to add a new field to an object. From the results we can rank the operations as:
- Fastest - adding a new entry to a dictionary using the indexing and assignment operators
- Faster - adding a new field to an object
- Slowest - updating a dictionary with a new entry

Overally, as far as time is concerned, we can tell that dictionary operations are faster than object operations.

## Space Tests

In [19]:
#create a job object
from jobs import Job
job_object = Job("Loans Manager","Mombasa","Full Time",\
        "KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
        "Accounting & Auditing")

#Create a job dictionary
job_dict = {"title":"Loans Manager","location":"Mombasa",\
"job_type":"Full Time","employer":"KUSCCO Limited (Kenya Union of Savings & Credit Co-operatives Limited)",\
"category":"Accounting & Auditing"}

#show that the internal values are the same
job_dict == job_object.__dict__

True

We will use `getsizeof()` to measure the size of our objects, and compare them. `getsizeof()` only measures the size of an object, excluding its fields. However, for dictionaries, it includes the size of the hash table.

In [20]:
from sys import getsizeof

In [21]:
#Size of Job class
getsizeof(Job)

888

In [22]:
#Size of original Job() object
getsizeof(Job())

56

In [23]:
#Size of job_object
getsizeof(job_object)

56

In [24]:
#Size of dict object
getsizeof(dict)

400

In [25]:
#Size of empty dictionary {}
getsizeof({})

240

In [26]:
#Size of job_dict
getsizeof(job_dict)

240

- The `dict` class has a smaller memory footprint (400) compared to the `Job` class (1184).
- The `job_object` has a smaller size (56) compared to `job_dict` (240)

### Adding one entry to the original dictionary

In [27]:
#Size of job_dict with one additional entry using update()
job_dict2 = job_dict.copy()
job_dict2.update({'salary':50000})
getsizeof(job_dict2)

368

In [28]:
#Size of job_dict with one additional entry using index and assignment operator
job_dict['salary']=50000
getsizeof(job_dict)

368

### Adding one extra field to the original Job class

In [29]:
from jobs import Job2

In [30]:
# Size of Job2 class - with one extra field
getsizeof(Job2)

1056

In [31]:
# Size of Job2 default object - with one extra field
getsizeof(Job2())

56

### Adding two extra fields to the original Job class

In [32]:
from jobs import Job3

In [33]:
# Size of Job3 class - with two extra fields
getsizeof(Job3)

1056

In [34]:
# Size of Job3 default object - with two extra fields
getsizeof(Job3())

56

From our numbers above, we can tell that objects tend to take up smaller memory spaces compared to dictionaries, despite the class definitions being rather large. Dictionaries appear to have a bigger memory fingerprint due to their hash tables.

## Conclusion
Dictionaries are a very handy data structure to master in Python. They are suitable for use with unordered data that relies on relations instead. Caution should however be practiced to ensure we do not use dictionaries in the wrong way and end up slowing down execution of our code.